### PR TITLE
V0.6.2

### PR DESCRIPTION
**Pull request by yramanchuk**
Implementation of custom Email entry controller

### CODE REVIEW

File: FirebaseAuthUI/FIRAuthPickerViewController.m [hunk 1]
```diff
@@ -134,8 +134,12 @@ - (void)viewDidLayoutSubviews {
...
 
 - (void)signInWithEmail {
-  UIViewController *controller =
-      [[FIREmailEntryViewController alloc] initWithAuthUI:self.authUI];
+  UIViewController *controller;
+  if ([self.authUI.delegate respondsToSelector:@selector(emailEntryViewControllerForAuthUI:)]) {
+    controller = [self.authUI.delegate emailEntryViewControllerForAuthUI:self.authUI];
+  } else {
+    controller = [[FIREmailEntryViewController alloc] initWithAuthUI:self];
+  }
```
Comment (morganchen12) [inline FirebaseAuthUI/FIRAuthPickerViewController.m]: Assert `controller != nil` after this line would be good, but not necessary

File: FirebaseAuthUI/FIRAuthUI.h [hunk 2]
```diff
@@ -127,10 +136,23 @@ typedef void (^FIRAuthUIResultCallback)(FIRUser *_Nullable user, NSError *_Nulla
...
+    @param error Optionally; if an error occurs during Firebase signout, upon return contains an
+    NSError object that describes the problem; is nil otherwise. If Firebase error occurs all providers
+    are not logged-out and sign-out should be retried.
+    @return @YES when the sign out request was successful. @NO otherwise.
+    @remarks Possible error codes:
+    - @c FIRAuthErrorCodeKeychainError Indicates an error occurred when accessing the keychain.
+    The @c NSLocalizedFailureReasonErrorKey field in the @c NSError.userInfo dictionary
+    will contain more information about the error encountered.
+ */
+- (BOOL)signOut:(NSError *_Nullable *_Nullable)error;
```
Comment (morganchen12) [inline FirebaseAuthUI/FIRAuthUI.h]: Consider renaming to `signOutWithError:(NSError *_Nullable *_Nullable)error`.

File: FirebaseAuthUI/FIRAuthUIBaseViewController.h [hunk 3]
```diff
@@ -104,6 +104,12 @@ typedef void (^FIRAuthUIAlertActionHandler)(void);
  */
 - (void)pushViewController:(UIViewController *)viewController;
 
+/** @fn onBackButtonPressed:
+    @brief Pops the view controller from navigation stack. If current controller is root 
+    works as @c cancelAuthorization
+ */
+- (void)onBackButtonPressed;
```
Comment (morganchen12) [inline FirebaseAuthUI/FIRAuthUIBaseViewController.h]: Consider following the UIKit pattern of having a sender and passing the button in along with the event.

```objc
- (void)onBackButtonPressed:(id)sender;
```

File: FirebaseAuthUI/FIREmailEntryViewController.m [hunk 4]
```diff
@@ -94,16 +93,19 @@ - (void)viewWillAppear:(BOOL)animated {
 }
 
 #pragma mark - Actions
-
```
Comment (morganchen12) [inline FirebaseAuthUI/FIREmailEntryViewController.m]: why delete this line?

File: samples/objc/FirebaseUI-demo-objc/Samples/Auth/FIRCustomEmailEntryViewController.m [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  //override action of default 'Next' button to use custom layout elements
+  self.navigationItem.rightBarButtonItem.target = self;
+  self.navigationItem.rightBarButtonItem.action = @selector(onBtnNext:);
+}
+
+- (IBAction)onBack:(id)sender {
+  [self onBackButtonPressed];
+}
+- (IBAction)onBtnNext:(id)sender {
```
Comment (morganchen12) [inline samples/objc/FirebaseUI-demo-objc/Samples/Auth/FIRCustomEmailEntryViewController.m]: Use full words or sentence fragments for selector names. `onNextButtonPressed:`

File: samples/objc/FirebaseUI-demo-objc/Samples/Auth/FIRCustomEmailEntryViewController.m [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+#import "FIRCustomEmailEntryViewController.h"
+
+@interface FIRCustomEmailEntryViewController () <UITextFieldDelegate>
+@property (weak, nonatomic) IBOutlet UITextField *emailField;
+@property (weak, nonatomic) IBOutlet UIBarButtonItem *btnNext;
```
Comment (morganchen12) [inline samples/objc/FirebaseUI-demo-objc/Samples/Auth/FIRCustomEmailEntryViewController.m]: Use full words for property names. `nextButton`

File: FirebaseAuthUI/FIRAuthUIBaseViewController.h (hunk 3 above)
Comment (yramanchuk) [inline FirebaseAuthUI/FIRAuthUIBaseViewController.h]: I think developers may be hesitating what is class type of Button (UIButton/UINavItem or they even can use voice command) and if it's eligible to pass it.
We are not using this value, so I think it's better just to rename to something like this:
```
- (void)onBack
- (void)navigateBack
- (void)goBack
```

What do you think?

File: FirebaseAuthUI/FIRAuthPickerViewController.m (hunk 1 above)
Comment (yramanchuk) [inline FirebaseAuthUI/FIRAuthPickerViewController.m]: We already have same situation with authViewController. I guess this check is redundant (different code style over project)

Comment (morganchen12) [inline FirebaseAuthUI/FIRAuthPickerViewController.m]: 👍 I'm fine with that. I think the UIKit method we call into likely asserts anyway.

**Comment by yramanchuk** [comment]
all changes were applied

**Pull request merged**